narwhal/executor/src/lib.rs

```diff
@@ -120,20 +122,40 @@ pub async fn get_restored_consensus_output<State: ExecutionState>(
         .read_last_consensus_index()
         .map_err(SubscriberError::StoreError)?;
 
-    let next_cert_index = execution_state
+    // Execution state always keeps the index of the latest certificate that has been executed.
+    // However, in consensus_store the committed certificates are stored alongside the "next" index
+    // that should be assigned to a certificate. Thus, to successfully recover the un-executed
+    // certificates from the consensus_store we are incrementing the `next_certificate_index` by 1
+    // to align the semantics.
+    // TODO: https://github.com/MystenLabs/sui/issues/5819
+    let last_executed_index = execution_state
         .load_execution_indices()
         .await
-        .next_certificate_index;
+        .next_certificate_index
+        + 1;
 
-    if next_cert_index < consensus_next_index {
-        let missing = consensus_store
-            .read_sequenced_certificates(&(next_cert_index..=consensus_next_index - 1))?
-            .iter()
-            .zip(next_cert_index..consensus_next_index)
-            .filter_map(|(c, seq)| c.map(|digest| (digest, seq)))
-            .collect::<Vec<(CertificateDigest, SequenceNumber)>>();
+    debug!(
+        "Recovering with last executed index:{}, last consensus index:{}",
+        last_executed_index, consensus_next_index
+    );
 
-        for (cert_digest, seq) in missing {
+    // We use <= check here as we always want to recover at least the last committed certificate
+    // since we can't know whether the execution has been interrupted and there are still batches/transactions
+    // that need to be send for execution.
+    if last_executed_index <= consensus_next_index {
+        let missing = consensus_store
+            .read_sequenced_certificates(&(last_executed_index..=consensus_next_index))?
+            .into_iter()
+            .map(|(seq, digest)| (seq - 1, digest))
```

Comment:
why do we decrement the sequence?

Reply:
Exactly. The returned indexes from the `read_sequenced_certificates` are ahead +1 . So in order to align them to what we are sending to the executor we should do the decrement